File: src/org/ollide/java2smali/GenerateAction.java
package org.ollide.java2smali;

import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.LangDataKeys;
import com.intellij.openapi.compiler.CompilerManager;
import com.intellij.openapi.module.Module;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.roots.ProjectRootManager;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.PsiClassOwner;
import com.intellij.psi.PsiFile;
import com.intellij.psi.PsiManager;

public class GenerateAction extends AnAction {

    private static final String JAVA = "java";
    private static final String KOTLIN = "kt";

    public void actionPerformed(AnActionEvent e) {
        VirtualFile vFile = getVirtualFileFromContext(e);

        Project p = e.getProject();
        Module module = ProjectRootManager.getInstance(p).getFileIndex().getModuleForFile(vFile);
        PsiClassOwner file = (PsiClassOwner) PsiManager.getInstance(p).findFile(vFile);

        // Compile the vFile's module
        CompilerCallback compilerCallback = new CompilerCallback(module, file);
        CompilerManager.getInstance(p).compile(module, compilerCallback);
    }

    @Override
    public void update(AnActionEvent e) {
        boolean enabled = false;

        VirtualFile vFile = getVirtualFileFromContext(e);
        if (vFile != null) {
            String extension = vFile.getFileType().getDefaultExtension();
            Module m = ProjectRootManager.getInstance(e.getProject()).getFileIndex().getModuleForFile(vFile);
            enabled = (JAVA.equals(extension) || KOTLIN.equals(extension)) && m != null;
        }
        e.getPresentation().setEnabled(enabled);
    }

    private VirtualFile getVirtualFileFromContext(AnActionEvent e) {
        PsiFile psiFile = e.getData(LangDataKeys.PSI_FILE);
        if (psiFile == null) {
            return null;
        }
        return psiFile.getVirtualFile();
    }

}

File: src/main/kotlin/org/ollide/java2smali/GenerateAction.kt
package org.ollide.java2smali

import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.LangDataKeys
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.roots.ProjectRootManager
import com.intellij.openapi.vfs.VirtualFile

class GenerateAction : AnAction() {

    override fun actionPerformed(e: AnActionEvent) {
        LOG.debug("Action performed.")

        val vFile = getVirtualFileFromEvent(e) ?: return
        val project = e.project!!
        val module = ProjectRootManager.getInstance(project).fileIndex.getModuleForFile(vFile)!!

        DexCompiler(vFile, project, module).run()
    }

    override fun update(e: AnActionEvent) {
        var enabled = false

        getVirtualFileFromEvent(e)?.let {
            e.project?.let { project ->
                val m = ProjectRootManager.getInstance(project).fileIndex.getModuleForFile(it)
                val extension = it.fileType.defaultExtension
                enabled = (JAVA == extension || KOTLIN == extension) && m != null
            }
        }
        e.presentation.isEnabled = enabled
    }

    private fun getVirtualFileFromEvent(e: AnActionEvent): VirtualFile? {
        val psiFile = e.getData(LangDataKeys.PSI_FILE) ?: return null
        return psiFile.virtualFile
    }

    companion object {
        private val LOG = Logger.getInstance(GenerateAction::class.java)

        private const val JAVA = "java"
        private const val KOTLIN = "kt"
    }

}
